Modules/_sqlite/connection.c

```diff
@@ -1442,6 +1457,173 @@ pysqlite_connection_iterdump(pysqlite_Connection* self, PyObject* args)
+    bckhandle = sqlite3_backup_init(bckconn, "main", self->db, name);
+    Py_END_ALLOW_THREADS
+
+    if (bckhandle) {
+        do {
+            Py_BEGIN_ALLOW_THREADS
+            rc = sqlite3_backup_step(bckhandle, pages);
+            Py_END_ALLOW_THREADS
+
+            if (progress != Py_None) {
+                if (!PyObject_CallFunction(progress, "iii", rc,
+                                           sqlite3_backup_remaining(bckhandle),
+                                           sqlite3_backup_pagecount(bckhandle))) {
+                    /* User's callback raised an error: interrupt the loop and
+                       propagate it. */
+                    cberr = 1;
+                    rc = -1;
+                }
+            }
+
+            /* Sleep for a while if there are still further pages to copy and
+               the engine could not make any progress */
+            if (rc == SQLITE_BUSY || rc == SQLITE_LOCKED) {
+                Py_BEGIN_ALLOW_THREADS
+                sqlite3_sleep(sleep);
+                Py_END_ALLOW_THREADS
+            }
+        } while (rc == SQLITE_OK || rc == SQLITE_BUSY || rc == SQLITE_LOCKED);
+
+        Py_BEGIN_ALLOW_THREADS
+        rc = sqlite3_backup_finish(bckhandle);
+        Py_END_ALLOW_THREADS
+    } else {
+        rc = _pysqlite_seterror(bckconn, NULL);
+    }
+
+    if (cberr == 0 && rc != SQLITE_OK) {
+        /* We cannot use _pysqlite_seterror() here because the backup APIs do
+           not set the error status on the connection object, but rather on
+           the backup handle. */
+        if (rc == SQLITE_NOMEM) {
+            (void)PyErr_NoMemory();
+        } else {
+            PyErr_SetString(pysqlite_OperationalError, sqlite3_errstr(rc));
+        }
+    }
+
+    if (filename != NULL) {
+        Py_BEGIN_ALLOW_THREADS
+        SQLITE3_CLOSE(bckconn);
+        Py_END_ALLOW_THREADS
+    }
+
+    if (cberr == 0 && rc == SQLITE_OK) {
+        Py_INCREF(Py_None);
+        retval = Py_None;
+    } else {
+        /* Remove the probably incomplete/invalid backup */
+        if (filename != NULL) {
+            PyObject *exc, *val, *tb;
+            PyErr_Fetch(&exc, &val, &tb);
+
+#ifndef MS_WINDOWS
+            if (unlink(filename) < 0) {
+              (void) PyErr_SetFromErrnoWithFilenameObject(PyExc_OSError, target);
+            }
+#else
+            Py_ssize_t size;
+            wchar_t *wide = PyUnicode_AsWideCharString(target, &size);
+
+            if (wide != NULL) {
+                if (!DeleteFileW(wide)) {
+                    PyMem_Free(wide);
+                    (void) PyErr_SetExcFromWindowsErrWithFilenameObject(PyExc_OSError,
+                                                                        GetLastError(),
+                                                                        target);
+                }
+                else PyMem_Free(wide);
+            }
+#endif
+
+            _PyErr_ChainExceptions(exc, val, tb);
+        }
+    }
+
+finally:
```

Comment:
You have defined a finally target but don't use it to clean up resources. Traditionally Python's C either uses goto to perform cleanups or returns immediately.